### PR TITLE
feat: TwingateConnector object (auto provision and update)

### LINKED ISSUE #22: Feature Proposal: TwingateConnector Object
Allow defining a `TwingateConnector` that'll 
- Provision a connector via API - get access&refresh tokens
- Start a pod with said connector
- Monitor pod and allow auto updating it

This is a basic building block on top of which we could later define an object that creates multiple of these (like a `Deployment` with `replicas` prop)

## Technical implementation

Proposal for CRD:
```
apiVersion: twingate.com/v1beta
kind: TwingateConnector
metadata:
  name: my-connector
spec:
  name: My K8S Connector
  imagePolicy:
    schedule: "0 2 * * *"
    version: 0.1.x
  containerExtra:
    resources:
      requests:
        cpu: 314m
        memory: 128Mi
      limits:
        cpu: 500m
        memory: 128Mi
  podExtra: {}
```

On create operator will:
1. Provision a connector and get tokens
2. Create a child `Secret` and `Pod` objects
3. Monitor pod and secret and not allow changing them
4. Create timers based on `imagePolicy.check` that will
  1. Check for newer tag on docker hub (based on `imagePolicy.version`)
  2. Patch the pod's image if a newer tag is detected

### PR DESCRIPTION
**Pull request by ekampf**
## Related Tickets & Documents

Resolved #22 

## Changes

- Added CRD for `TwingateConnector`
- Added GraphQL API client support to get, create, delete and provision connectors 

## Notes

Example object (not that `image` and `imagePolicy` are mutually exclusive, can't use both):
```
apiVersion: twingate.com/v1beta
kind: TwingateConnector
metadata:
  name: my-connector
spec:
  image:
    repository: twingate/connector
    tag: 1.60.0
  imagePolicy:
    schedule: "0 2 * * *"
    version: "^1.0.0"
  containerExtra:
    resources:
      requests:
        cpu: 314m
        memory: 128Mi
      limits:
        cpu: 500m
        memory: 128Mi
  podExtra: {}
```

It can even be as minimal as:
```
apiVersion: twingate.com/v1beta
kind: TwingateConnector
metadata:
  name: my-connector
spec: {}
```

In which case the operator will default to `twingate/connector:1` image and server will give the connector its name.

On creation the object:
- Call Twingate API and create connector, generate tokens for it
- Create two child objects in K8s:
  - `Secret` to hold the tokens
  - `Pod` that actually runs the connector
- if there's an `imagePolicy` - Annotate the connector object on the next check schedule

Separately, the operator runs a timer every `1m` and checks all connectors with the annotation:
  - fetch all image versions from dockerhub
  - filter by the version spec (we support npm specification)
  - get the latest and patches the pod. (if there's any change from current state kuberenets will reconciliate and restart the pod)
  - re-annotate the connector object with next check datetime

### Note for review

Suggested order:
1. All the aux code:
    1. The new CRD - `deploy/twingate-operator/crds/twingate.com.twingateconnector.yaml`
    2. The Python CRD module - `app/crds.py`
    3. The `app/version_policy_providers` module
    4. API Changes to support connector API calls - `app/api/client_connectors.py`
2. The actual handler logic at `app/handlers/handlers_connectors.py`

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import os
+
+import pytest
+
+
+def pytest_collection_modifyitems(session, config, items):
```
Comment (ekampf) [inline conftest.py:6]: This just moved from `tests_integration/conftest.py` here to the root.  Its a better fit...

File: tests_integration/conftest.py [hunk 2]
```diff
@@ -32,10 +32,3 @@ def kopf_settings():
 @pytest.fixture()
 def unique_resource_name(request):
     return request.node.name.replace("_", "-") + "-" + str(uuid.uuid4())
-
-
-def pytest_collection_modifyitems(session, config, items):
```
Comment (ekampf) [inline tests_integration/conftest.py:37]: Moved to root `conftest.py`

File: app/handlers/handlers_connectors.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+@kopf.on.update(
+    "twingateconnector", labels={"twingate.com/connector-pod-deleted": "true"}
+)
+def twingate_connector_recreate_pod(body, namespace, memo, logger, **_):
+    """Recreates the Connector's Pod.
+
+    When pod is deleted we can't recreate it right away because we want to
+    use the same name. So when it's deleted, `twingate_connector_pod_deleted` annotates
+    it's connector object so that we get heere and can recreate it.
```
Comment (minhtule) [inline app/handlers/handlers_connectors.py:182]: nit: should be "its" and "here" instead of "heere"?

File: app/handlers/handlers_connectors.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    logger.info("Got a delete request: %s. Status: %s", spec, status)
+    if not status:
+        return
+
+    if connector_id := spec.get("id"):
+        logger.info("Deleting connector %s", connector_id)
+        memo.twingate_client.connector_delete(connector_id)
+
+    try:
+        # Remove label from pod so it's delete handler isn't triggered
```
Comment (minhtule) [inline app/handlers/handlers_connectors.py:208]: nit: should be "its delete handler"

File: tests_integration/test_connector_flows.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        # Secret was properly created
+        assert secret["data"] == {"TWINGATE_ACCESS_TOKEN": ANY, "TWINGATE_REFRESH_TOKEN": ANY}  # fmt: skip
+        # pod was properly created
+        assert pod["status"]["phase"] == "Running"
+        assert pod["metadata"]["annotations"]["twingate.com/kopf-managed"] == "yes"
+        assert pod["metadata"]["labels"]["twingate.com/connector"] == connector_name
+        assert pod["metadata"]["ownerReferences"][0]["name"] == connector_name
+        assert pod["metadata"]["ownerReferences"][0]["kind"] == "TwingateConnector"
+
+        kubectl_delete(f"tc/{connector_name}")
```
Comment (minhtule) [inline tests_integration/test_connector_flows.py:52]: Should we also test that deleting the pod and assert that it's recreated correctly? 🤔

File: tests_integration/test_crds_connector.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                """
+            )
+
+    def test_name_is_immutable(self, unique_connector_name):
+        result = kubectl_create(
+            f"""
+            apiVersion: twingate.com/v1beta
+            kind: TwingateConnector
+            metadata:
+                name: {unique_connector_name}
```
[the remark this answers is not in the document]
Comment (ekampf) [reply tests_integration/test_crds_connector.py:155]: `metadata.name` is required by k8s

File: app/handlers/handlers_connectors.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    kopf.label(pod, {"twingate.com/connector": crd.metadata.name})
+
+    kapi = kubernetes.client.CoreV1Api()
+    kapi.create_namespaced_pod(namespace=namespace, body=pod)
+
+
+@kopf.on.delete("twingateconnector")
+def twingate_connector_delete(spec, meta, status, namespace, memo, logger, **kwargs):
+    logger.info("Got a delete request: %s. Status: %s", spec, status)
+    if not status:
```
[the remark this answers is not in the document]
Comment (ekampf) [reply app/handlers/handlers_connectors.py:200]: If there'sd no `status` yet then its a new object that even `create` event didnt run on yet - so its definately not synced to Twingate and we should allow deleting it and do nothing

File: app/handlers/handlers_connectors.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+
+def get_connector_secret(
+    access_token: str, refresh_token: str
+) -> kubernetes.client.V1Secret:
+    return kubernetes.client.V1Secret(
+        immutable=True,
+        data={
+            "TWINGATE_ACCESS_TOKEN": base64.b64encode(
+                access_token.encode("ascii")
```
[the remark this answers is not in the document]
Comment (ekampf) [reply app/handlers/handlers_connectors.py:74]: Out token is a JWT - so 3 base64 values. Can a base64 string have non ascii values?

[the remark this answers is not in the document]
Comment (minhtule) [reply app/handlers/handlers_connectors.py:75]: I see. Then should we also use `ascii` when decoding the encoded base64 in the next line?

File: app/handlers/handlers_connectors.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+
+    # fmt: off
+    pod_spec = {
+        "containers": [
+            {
+                "env": [
+                    {"name": "TWINGATE_LABEL_DEPLOYED_BY", "value": "operator"},
+                    {"name": "TWINGATE_LABEL_OPERATOR_VERSION", "value": get_version()},
+                    {"name": "TWINGATE_URL", "value": tenant_url},
+                    {"name": "TWINGATE_LOG_LEVEL", "value": "7"},
```
Comment (liorr) [inline app/handlers/handlers_connectors.py:43]: We should pass this a parameter like `tenant_url`

File: app/handlers/handlers_connectors.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (liorr) [reply app/handlers/handlers_connectors.py:75]: instead of encoding and decoding just pass `string_data` instead of `data`

**Review by liorr — COMMENTED** [review]
The first pass, I have a couple of comments. Please check.

File: app/handlers/handlers_connectors.py (hunk 9 above)
Comment (ekampf) [reply app/handlers/handlers_connectors.py:42]: @liorr yes its on my list of things I want but on separate PR...  wanted to keep this Pr to minimum as its already big

**Review by liorr — COMMENTED** [review]
Please review my comments

File: scripts/minikube-smoketests.sh [hunk 10]
```diff
@@ -36,7 +36,7 @@ main() {
 	install_minikube_release
 
 	# create a cluster
-	"${MINIKUBE}" start --force
+	"${MINIKUBE}" start --force --kubernetes-version=1.27.4
```
[the remark this answers is not in the document]
Comment (ekampf) [reply scripts/minikube-smoketests.sh:39]: @liorr its a tempo fix - minikube used to have rbac allow-all setting by default and with latest update it does not.
so on a separate PR I need to fix this - remove the version loimitation and apply a yaml that gives default role access to everything

File: app/handlers/handlers_connectors.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    crd = TwingateConnectorCRD(**body)
+    if not crd.spec.image_policy:
+        patch.meta["annotations"] = {ANNOTATION_NEXT_VERSION_CHECK: None}
+        return
+
+    now = pendulum.now("UTC").start_of("minute")
+    next_check = pendulum.parse(
+        body["metadata"]["annotations"][ANNOTATION_NEXT_VERSION_CHECK]
+    )
+    if now < next_check:
```
[the remark this answers is not in the document]
Comment (ekampf) [reply app/handlers/handlers_connectors.py:149]: See comment re `when` filter above...

File: scripts/minikube-smoketests.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (ekampf) [reply scripts/minikube-smoketests.sh]: turns out its fixed on minikube's end and this is no longer needed

**Review by liorr — APPROVED** [review]
LGTM

**Pull request merged**